### PR TITLE
Mention processing: handle bracketed and possessive mentions

### PR DESCRIPTION
**Pull request by sporksmith**
Handle possessive mentions (e.g. "Did you get JCDenton's report?") and bracketed/quoted mentions (e.g. "Hey, you missed: \<JCDenton\> it's a bomb!").

Signed-off-by: Jim Newsome <jnewsome@torproject.org>

### CODE REVIEW

**Comment by sporksmith** [comment]
Debugging the failing tests via the GH action.

Btw I wasn't able to get them running locally due to missing dependencies, but I'm not very familiar with node etc so may be missing something. I tried `npm install package.json` and `npm install promiseutil` but am still getting:

```
$ npm run check

> matrix-appservice-irc@0.25.0 check /home/jnewsome/projects/matrix-appservice-irc
> npm test && npm run lint


> matrix-appservice-irc@0.25.0 test /home/jnewsome/projects/matrix-appservice-irc
> BLUEBIRD_DEBUG=1 node --max_old_space_size=3072 node_modules/jasmine/bin/jasmine.js --stop-on-failure=true

{ Error: Cannot find module '../../lib/promiseutil'
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:636:15)
    at Function.Module._load (internal/modules/cjs/loader.js:562:25)
    at Module.require (internal/modules/cjs/loader.js:692:17)
    at require (internal/modules/cjs/helpers.js:25:18)
    at Object.<anonymous> (/home/jnewsome/projects/matrix-appservice-irc/spec/util/irc-client-mock.js:8:15)
    at Module._compile (internal/modules/cjs/loader.js:778:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:789:10)
    at Module.load (internal/modules/cjs/loader.js:653:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:593:12)
    at Function.Module._load (internal/modules/cjs/loader.js:585:3)
    at Module.require (internal/modules/cjs/loader.js:692:17)
    at require (internal/modules/cjs/helpers.js:25:18)
    at Object.<anonymous> (/home/jnewsome/projects/matrix-appservice-irc/spec/util/test.js:8:17)
    at Module._compile (internal/modules/cjs/loader.js:778:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:789:10)
    at Module.load (internal/modules/cjs/loader.js:653:32) code: 'MODULE_NOT_FOUND' }
npm ERR! Test failed.  See above for more details.
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! matrix-appservice-irc@0.25.0 check: `npm test && npm run lint`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the matrix-appservice-irc@0.25.0 check script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/jnewsome/.npm/_logs/2021-05-02T18_09_37_094Z-debug.log

```

**Comment by sporksmith** [comment]
Ok, I've gotten the unit tests passing again. Ended up scaling back the change to more specifically handle the use-cases I've come across. I do still think it might be worth refactoring to use \b, and maybe doing secondary checks to see if it's inside a URL etc, but that's more than I can bite off at the moment.

AFAIK the tests that are still failing are unrelated to my change, but LMK if that's not the case and how to go about fixing it.

Would also still appreciate pointers for getting the tests running locally, since I'd like to continue contributing. I'm a member of the Tor Project and we just started bridging our IRC channels to Matrix; would be happy to continue fixing small bugs we stumble across.

File: changelog.d/1303.bugfix [hunk 1]
```diff
@@ -0,0 +1 @@
+Also detect username mentions bounded by ',<,>,&
```
Comment (Half-Shot) [inline changelog.d/1303.bugfix:1]: Just so it's descriptive to the user :)

```suggestion
Detect IRC username mentions bounded by ',<,> or &
```

**Review by Half-Shot — APPROVED** [review]
I think this generally looks sane to me. Have just mentioned that we should probably be a little more descriptive on the changelog.

**Comment by Half-Shot** [comment]
>Would also still appreciate pointers for getting the tests running locally, since I'd like to continue contributing. I'm a member of the Tor Project and we just started bridging our IRC channels to Matrix; would be happy to continue fixing small bugs we stumble across.

Hiya, yeah it should just be a case of `npm i && npm run test` really. You need to run `npm run build` each time you change the code as we compile the typescript down to JavaScript and test the JS code.

**Comment by sporksmith** [comment]
> Hiya, yeah it should just be a case of `npm i && npm run test` really. You need to run `npm run build` each time you change the code as we compile the typescript down to JavaScript and test the JS code.

@Half-Shot Thanks, but still no luck :/

```
$ npm run build && npm i && npm run test

> matrix-appservice-irc@0.25.0 build /home/jnewsome/projects/matrix-appservice-irc
> tsc --project ./tsconfig.json


> matrix-appservice-irc@0.25.0 prepare /home/jnewsome/projects/matrix-appservice-irc
> npm run build


> matrix-appservice-irc@0.25.0 build
> tsc --project ./tsconfig.json

npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@2.3.1 (node_modules/fsevents):
npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@2.3.1: wanted {"os":"darwin","arch":"any"} (current: {"os":"linux","arch":"x64"})

audited 699 packages in 4.001s

44 packages are looking for funding
  run `npm fund` for details

found 1 high severity vulnerability
  run `npm audit fix` to fix them, or `npm audit` for details

> matrix-appservice-irc@0.25.0 test /home/jnewsome/projects/matrix-appservice-irc
> BLUEBIRD_DEBUG=1 node --max_old_space_size=3072 node_modules/jasmine/bin/jasmine.js --stop-on-failure=true

{ Error: Cannot find module 'worker_threads'
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:636:15)
    at Function.Module._load (internal/modules/cjs/loader.js:562:25)
    at Proxyquire._require (/home/jnewsome/projects/matrix-appservice-irc/node_modules/proxyquire/lib/proxyquire.js:166:19)
    at require (internal/modules/cjs/helpers.js:25:18)
    at Object.<anonymous> (/home/jnewsome/projects/matrix-appservice-irc/lib/workers/MetricsWorker.js:7:26)
    at Module._compile (internal/modules/cjs/loader.js:778:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:789:10)
    at Object.require.extensions.(anonymous function) (/home/jnewsome/projects/matrix-appservice-irc/node_modules/proxyquire/lib/proxyquire.js:276:43)
    at Module.load (internal/modules/cjs/loader.js:653:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:593:12)
    at Function.Module._load (internal/modules/cjs/loader.js:585:3)
    at Proxyquire._require (/home/jnewsome/projects/matrix-appservice-irc/node_modules/proxyquire/lib/proxyquire.js:166:19)
    at require (internal/modules/cjs/helpers.js:25:18)
    at Object.<anonymous> (/home/jnewsome/projects/matrix-appservice-irc/lib/bridge/IrcBridge.js:47:25)
    at Module._compile (internal/modules/cjs/loader.js:778:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:789:10)
    at Object.require.extensions.(anonymous function) (/home/jnewsome/projects/matrix-appservice-irc/node_modules/proxyquire/lib/proxyquire.js:276:43)
    at Module.load (internal/modules/cjs/loader.js:653:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:593:12)
    at Function.Module._load (internal/modules/cjs/loader.js:585:3)
    at Proxyquire._require (/home/jnewsome/projects/matrix-appservice-irc/node_modules/proxyquire/lib/proxyquire.js:166:19)
    at require (internal/modules/cjs/helpers.js:25:18) code: 'MODULE_NOT_FOUND' }
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! matrix-appservice-irc@0.25.0 test: `BLUEBIRD_DEBUG=1 node --max_old_space_size=3072 node_modules/jasmine/bin/jasmine.js --stop-on-failure=true`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the matrix-appservice-irc@0.25.0 test script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/jnewsome/.npm/_logs/2021-05-06T14_36_43_980Z-debug.log
```

**Comment by Half-Shot** [comment]
That looks like an issue with Node 10 as it doesn't support worker threads, are you able to try building on a newer version of node. If that doesn't work, we might be able to force Node 10 to work by telling it to use experimental worker threads as suggested in https://matrix-org.github.io/matrix-appservice-irc/latest/bridge_setup

**Pull request merged**